### ISSUE TITLE
Memory consumption is huge when using einsum() with a non intermediate tensor path

### ISSUE THREAD

**Issue by LUOXIAO92** [cupy/cupy#7180]
### Description

What happened: Memory consumption is huge when using einsum() with a non intermediate tensor path
What you expected to happen: Memory consumption should be reasonable with a non intermediate tensor path

When I use the non intermediate tensor path
```
path = ['einsum_path', (0, 1, 2, 3)]
```
to calculate 
```
cp.einsum("acke,bdel,abi,cdj->ijkl", T, T, U, np.conj(U), optimize=path)
```
and then got the following error. The reasonable memory consumption should be about $O(n^4)$, but when I set $D=30$ in the sample code, the memory consumption becomes very large shown as follows. However, in the case of $D=30$ when use the path including intermediate tensor, 
```
path = ['einsum_path', (0, 2), (0, 1), (0, 1)]
```
memory consumption becomes reasonable.
```
Traceback (most recent call last):
  File "/work/0/TRG/luoxiao/test/cotengra/einsum_test.py", line 14, in <module>
    Timp0 = cp.einsum("acke,bdel,abi,cdj->ijkl", T, T, U, np.conj(U), optimize=path)
  File "/home/.miniconda3/envs/py3.9/lib/python3.9/site-packages/cupy/linalg/_einsum.py", line 676, in einsum
    arr_out, sub_out = reduced_binary_einsum(
  File "/home/.miniconda3/envs/py3.9/lib/python3.9/site-packages/cupy/linalg/_einsum.py", line 421, in reduced_binary_einsum
    arr_out = cupy.matmul(tmp0, tmp1).reshape(shapes_out)
  File "/home/.miniconda3/envs/py3.9/lib/python3.9/site-packages/cupy/_core/_gufuncs.py", line 671, in __call__
    outs = cupy.empty(out_shape, dtype=ret_dtype, order=filter_order)
  File "/home/.miniconda3/envs/py3.9/lib/python3.9/site-packages/cupy/_creation/basic.py", line 22, in empty
    return cupy.ndarray(shape, dtype, order=order)
  File "cupy/_core/core.pyx", line 136, in cupy._core.core.ndarray.__new__
  File "cupy/_core/core.pyx", line 224, in cupy._core.core._ndarray_base._init
  File "cupy/cuda/memory.pyx", line 742, in cupy.cuda.memory.alloc
  File "cupy/cuda/memory.pyx", line 1419, in cupy.cuda.memory.MemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 1440, in cupy.cuda.memory.MemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 1120, in cupy.cuda.memory.SingleDeviceMemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 1141, in cupy.cuda.memory.SingleDeviceMemoryPool._malloc
  File "cupy/cuda/memory.pyx", line 1379, in cupy.cuda.memory.SingleDeviceMemoryPool._try_malloc
cupy.cuda.memory.OutOfMemoryError: Out of memory allocating 11,664,000,000 bytes (allocated so far: 23,366,881,280 bytes).
```

### To Reproduce

```py
import cupy as cp
import numpy as np
import time

#path = ['einsum_path', (0, 2), (0, 1), (0, 1)]
path = ['einsum_path', (0, 1, 2, 3)]
D=30
T = cp.random.random((D,D,D,D)) + 1j*cp.random.random((D,D,D,D))
U = cp.random.random((D,D,D))   + 1j*cp.random.random((D,D,D))
t0=time.time()
#Timp0 = np.einsum("acke,bdel,abi,cdj->ijkl", T, T, U, np.conj(U), optimize=path)
Timp0 = cp.einsum("acke,bdel,abi,cdj->ijkl", T, T, U, np.conj(U), optimize=path)
t1=time.time()
print("{:.2e} s".format(t1-t0))
```

### Installation

Conda-Forge (`conda install ...`)

### Environment

```
OS                           : Linux-3.10.0-1160.59.1.el7.x86_64-x86_64-with-glibc2.17
Python Version               : 3.9.12
CuPy Version                 : 11.0.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.22.3
SciPy Version                : 1.7.3
Cython Build Version         : 0.29.30
Cython Runtime Version       : None
CUDA Root                    : /home/.miniconda3/envs/py3.9
nvcc PATH                    : None
CUDA Build Version           : 11020
CUDA Driver Version          : 11060
CUDA Runtime Version         : 11060
cuBLAS Version               : (available)
cuFFT Version                : 10600
cuRAND Version               : 10209
cuSOLVER Version             : (11, 3, 2)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 6)
Thrust Version               : 101000
CUB Build Version            : 101000
Jitify Build Version         : 3c4a4ba
cuDNN Build Version          : None
cuDNN Version                : None
NCCL Build Version           : 21212
NCCL Runtime Version         : 21207
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : Tesla V100-PCIE-32GB
Device 0 Compute Capability  : 70
Device 0 PCI Bus ID          : 0000:1C:00.0
Device 1 Name                : Tesla V100-PCIE-32GB
Device 1 Compute Capability  : 70
Device 1 PCI Bus ID          : 0000:1D:00.0
Device 2 Name                : Tesla V100-PCIE-32GB
Device 2 Compute Capability  : 70
Device 2 PCI Bus ID          : 0000:3F:00.0
Device 3 Name                : Tesla V100-PCIE-32GB
Device 3 Compute Capability  : 70
Device 3 PCI Bus ID          : 0000:40:00.0
```

**Comment by emcastillo** [comment]
Can you try cutensornet with cuquantum? this should give much better results.
It will require cuquantum-python >= 22.03 and cuTENSOR >= 1.5.0


Setting the following env var will activate it
```
CUPY_ACCELERATORS=cutensornet
```

**Comment by leofang** [comment]
> and doesn't work with CuPy

I believe this refers to the cuTENSOR 1.6.0 bug (https://github.com/cupy/cupy/issues/6927) that is fixed in cuTENSOR 1.6.1. We have not pushed the cuTENSOR 1.6.1 wheel yet, as we plan to re-structure/rename the wheel. Hopefully it can be done later this month.

In the meanwhile, you can just do `pip install cuquantum-python "cutensor==1.5.*"` to stay on cuTENSOR 1.5.0.

**Comment by LUOXIAO92** [comment]
> Can you try cutensornet with cuquantum? this should give much better results. It will require cuquantum-python >= 22.03 and cuTENSOR >= 1.5.0 (current cuTensor in PyPi is not updated and doesn't work with CuPy, so you will need to manually install it using:
> 
> ```
> python -m cupyx.tools.install_library --cuda 11.x --library cutensor
> ```
> 
> Setting the following env var will activate it
> 
> ```
> CUPY_ACCELERATORS=cutensornet
> ```

thanks, cuquantun's einsum() works without specify the path in the conctraction of the large bound dimension of tensor. But when I specify the following path
```python
optimize=[(0, 1, 2, 3)]
```
the following error occurs.
```
TypeError: The path must be a sequence of pairs in the linear format accepted by numpy.einsum_path.
```
And moreover, when I use mpiexec -np  to run the program independently by setting
```python
from mpi4py import MPI
comm = MPI.COMM_WORLD 
myrank = comm.Get_rank() 
cuda = cp.cuda.Device(myrank)
cuda.use()
```
performancewarning and valueerror occurs.
```
The device where the array resides (2) is different from the current device (0).
The device where the array resides (1) is different from the current device (0).
The device where the array resides (3) is different from the current device (0).
ValueError: All tensors in the network are not on the same device. Devices = {0, 2}.
ValueError: All tensors in the network are not on the same device. Devices = {0, 1}.
ValueError: All tensors in the network are not on the same device. Devices = {0, 3}.
```
And only gpu0 is working
```
$ nvidia-smi
Wed Nov  9 22:48:42 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-PCIE...  On   | 00000000:1C:00.0 Off |                    0 |
| N/A   67C    P0   253W / 250W |  17574MiB / 32768MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-PCIE...  On   | 00000000:1D:00.0 Off |                    0 |
| N/A   41C    P0    38W / 250W |   7977MiB / 32768MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-PCIE...  On   | 00000000:3F:00.0 Off |                    0 |
| N/A   40C    P0    38W / 250W |   7977MiB / 32768MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-PCIE...  On   | 00000000:40:00.0 Off |                    0 |
| N/A   39C    P0    37W / 250W |   7977MiB / 32768MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A    119226      C   python3                         12111MiB |
|    0   N/A  N/A    119227      C   python3                          1819MiB |
|    0   N/A  N/A    119228      C   python3                          1819MiB |
|    0   N/A  N/A    119229      C   python3                          1819MiB |
|    1   N/A  N/A    119227      C   python3                          7973MiB |
|    2   N/A  N/A    119228      C   python3                          7973MiB |
|    3   N/A  N/A    119229      C   python3                          7973MiB |
+-----------------------------------------------------------------------------+
```

**Comment by leofang** [comment]
Sorry I dropped the ball...

> But when I specify the following path
> 
> ```python
> optimize=[(0, 1, 2, 3)]
> ```
> 
> the following error occurs.

This is expected, cuQuantum/cuTensorNet only supports pairwise-contraction paths, a subset of what NumPy supports. In NumPy's case, it's contracting the whole thing in batch, which is not something straightforward to express with accelerators. If it's a trivial path, you can manually break it down as `[(0,1)]*3`, or use some helper functions to convert the path, or just let cuTensorNet compute the path for you.

> And moreover, when I use mpiexec -np to run the program independently by setting
> 
> ```python
> from mpi4py import MPI
> comm = MPI.COMM_WORLD 
> myrank = comm.Get_rank() 
> cuda = cp.cuda.Device(myrank)
> cuda.use()
> ```
> 
> performancewarning and valueerror occurs.

This is because cuTensorNet expects the `cuquantum.Network` object created on the same device as all of the input tensors, but calling `cuda.uses()` happens to not make older cuTensorNet aware of this. Starting [cuQuantum v22.11](https://docs.nvidia.com/cuda/cuquantum/python/release_notes.html#cuquantum-python-v22-11-0) this has been worked around:
> Fix a potential multi-device bug in the internal device context switch.

or if you can update to CuPy v12.0.0+ it should work too (without updating cuQuantum). This was addressed in #6965 on the CuPy side.

**Comment by leofang** [comment]
Closing due to inactivity. For questions about cuQuantum/cuTensorNet, please kindly open a new discussion thread in [the cuQuantum repo](https://github.com/NVIDIA/cuQuantum). Thanks.
